File: Python/ByTier/F/1821_F_Timber.py
# Problem: CF 1821 F - Timber
# https://codeforces.com/contest/1821/problem/F

"""
Code Purpose:
This code solves the problem of counting the number of "unfortunate" alleys with m trees of height k,
where each tree can fall either left or right, and no two trees can occupy the same spot when fallen.
The solution uses dynamic programming with combinatorics to calculate valid configurations.

Algorithms/Techniques:
- Combinatorial mathematics using precomputed factorials and inverse factorials for binomial coefficients
- Inclusion-exclusion principle to count valid arrangements
- Fast exponentiation for modular arithmetic
- Preprocessing of powers of 2 for efficiency

Time Complexity:
O(n) - where n is the number of spots for trees. The dominant factor is the precomputation of factorials and powers of 2.

Space Complexity:
O(n) - for storing precomputed factorial and inverse factorial arrays, and powers of 2 up to 10^6.
"""

mod = 998244353

# Precompute factorials and inverse factorials for combinations
table_size = 4 * 10**5
fac = [1] * (table_size + 1)
finv = [1] * (table_size + 1)

for i in range(2, table_size + 1):
    fac[i] = fac[i - 1] * i % mod
finv[table_size] = pow(fac[table_size], mod - 2, mod)
for i in range(table_size - 1, -1, -1):
    finv[i] = finv[i + 1] * (i + 1) % mod


def binom(n, k):
    # Compute binomial coefficient C(n, k) mod 998244353
    if n < 0 or k < 0:
        return 0
    if k > n:
        return 0
    global table_size, fac, finv
    if n > table_size:
        # Extend the tables if needed
        fac += [0] * (n - table_size)
        finv += [0] * (n - table_size)
        for i in range(table_size + 1, n + 1):
            fac[i] = fac[i - 1] * i % mod
        finv[n] = inv(fac[n])
        for i in range(n - 1, table_size, -1):
            finv[i] = finv[i + 1] * (i + 1) % mod
        table_size = n
    return (fac[n] * finv[k] % mod) * finv[n - k] % mod


def fpow(x, k):
    # Fast exponentiation: compute x^k mod mod
    res = 1
    while k:
        if k & 1:  # If k is odd
            res = res * x % mod
        x = x * x % mod
        k >>= 1  # Divide k by 2
    return res


def inv(a):
    # Compute modular inverse of a mod mod using Fermat's little theorem
    if a < table_size:
        return fac[a - 1] * finv[a] % mod
    return fpow(a, mod - 2)


def modint_to_frac(a):
    # This function is not used in the final computation but provides modular fraction representation
    a %= mod
    if a == 0:
        return "0/1"
    for X in range(1, 10000):
        Y = a * X % mod
        if 0 < Y < 10000:
            return str(Y) + "/" + str(X)
        if mod - 10000 < Y < mod:
            return "-" + str(mod - Y) + "/" + str(X)
    return "inexpressible"


# Read input
n, m, k = map(int, input().split())

# Precompute powers of 2 up to 10^6 for efficiency
pow2 = [1] * (10**6)
for i in range(1, 10**6):
    pow2[i] = pow2[i - 1] * 2 % mod

ans = 0
sgn = 1  # Sign for inclusion-exclusion principle

# Iterate over all possible arrangements using inclusion-exclusion
for i in range(m + 1):
    a = m - i  # Number of trees falling to the left
    b = i      # Number of trees falling to the right
    c = n - (a * (k + 1)) - b * (2 * k + 1)  # Remaining space after placing trees
    if c >= 0:
        # Add contribution to total count using binomial coefficients
        ans += sgn * binom(a + b + c, a) * binom(b + c, b) % mod * pow2[a]
        ans %= mod
    sgn = -sgn  # Alternate sign for inclusion-exclusion

# Output result
print(ans % mod)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1821/1821_F_Timber.cpp
// Problem: CF 1821 F - Timber
// https://codeforces.com/contest/1821/problem/F

/*
Code Purpose:
This solution calculates the number of "unfortunate" alleys where m trees of
height k can be chopped down such that no tree falls on the mall (index 0) or
road (index n+1), and each spot is covered by at most one tree. The algorithm
uses inclusion-exclusion principle combined with combinatorics to compute the
count.

Algorithms/Techniques:
- Inclusion-Exclusion Principle
- Combinatorial mathematics (binomial coefficients)
- Modular arithmetic and modular exponentiation
- Preprocessing factorials and their modular inverses

Time Complexity: O(n)
Space Complexity: O(n)

*/

#include <ctype.h>
#include <iostream>
#include <stdio.h>

#define m_p(a, b) make_pair(a, b)
#define pb(a) push_back(a)
#define ll long long
#define ull unsigned long long
#define inf 0x7FFFFFFF
#define inff 9223372036854775807
#define rep(i, l, r) for (int i = l; i < r; ++i)
#define repp(i, l, r) for (int i = l; i <= r; ++i)
#define per(i, r, l) for (int i = r - 1; i >= l; --i)
#define pii pair<int, int>
#define fi first
#define se second
#define p_q priority_queue
#define Kill(x)                                                                \
  {                                                                            \
    cout << x << endl;                                                         \
    return 0;                                                                  \
  }
#define all(x) x.begin(), x.end()
#define mod 998244353
#define int ll
using namespace std;
inline int rd() {
  int x(0), f(1);
  char ch = getchar();
  while (!isdigit(ch)) {
    if (ch == '-')
      f = -f;
    ch = getchar();
  }
  while (isdigit(ch)) {
    x = (x << 3) + (x << 1) + (ch ^ 48);
    ch = getchar();
  }
  return x * f;
}
inline void out(int X) {
  if (X < 0) {
    X = ~(X - 1);
    putchar('-');
  }
  if (X > 9)
    out(X / 10);
  putchar(X % 10 + '0');
}
ll pw(int x, int d) {
  if (!d)
    return 1;
  ll t = pw(x, d >> 1);
  t = t * t % mod;
  if (d & 1)
    return t * x % mod;
  return t;
}
#define MAX 300005
ll jie[MAX]; // Factorial array
ll inv[MAX]; // Modular inverse of factorial
ll pw2[MAX]; // Powers of 2 modulo mod
inline ll Comb(int x, int y) {
  if (y > x)
    return 0;
  return jie[x] * inv[x - y] % mod * inv[y] % mod;
}
signed main() {
  int n = rd();
  int m = rd();
  int k = rd();
  jie[0] = 1;
  rep(i, 1, MAX) jie[i] = (jie[i - 1] * i) % mod;
  inv[MAX - 1] = pw(jie[MAX - 1], mod - 2);
  per(i, MAX - 1, 0) inv[i] = (inv[i + 1] * (i + 1)) % mod;
  pw2[0] = 1;
  rep(i, 1, MAX) pw2[i] = (pw2[i - 1] << 1) % mod;
  int nw = 1;
  ll ans = 0;
  repp(i, 0, m) {
    ans = (ans +
           pw2[m - i] * nw % mod * Comb(n - m * k - i * k, m) % mod *
               Comb(m, i) % mod +
           mod) %
          mod;
    nw *= -1;
  }
  cout << ans << endl;
  return 0;
}

// https://github.com/VaHiX/CodeForces/